phys/module_sf_noahdrv.F

```diff
@@ -3740,7 +3740,49 @@ SUBROUTINE lsm_mosaic(DZ8W,QV3D,P8W3D,T3D,TSK,                      &
                      IVGTYP(I,J) == LCZ_6      .or. IVGTYP(I,J) == LCZ_7 .or. IVGTYP(I,J) == LCZ_8 .or. &
                      IVGTYP(I,J) == LCZ_9      .or. IVGTYP(I,J) == LCZ_10 .or. IVGTYP(I,J) == LCZ_11 ) THEN
 
-                  UTYPE_URB = UTYPE_URB2D(I,J) !urban type (low, high or industrial)
+                  !  UTYPE_URB = UTYPE_URB2D(I,J) !urban type (low, high or industrial)
+                  !  this need to be changed in the mosaic danli
+                  IF (slucm_distributed_drag) THEN
+                    IF (IVGTYP(I, J) == ISURBAN) THEN
+                      UTYPE_URB = 2
+                      FRC_URB2D(I, J) = MAX(0.1, MIN(0.9, 1 - shdavg(I, J) / 100.))
+                    END IF
+                  ELSE IF (use_wudapt_lcz == 1) THEN
+                    IF(IVGTYP(I,J)==ISURBAN) UTYPE_URB=5
+                    IF(IVGTYP(I,J)==LCZ_1) UTYPE_URB=1
+                    IF(IVGTYP(I,J)==LCZ_2) UTYPE_URB=2
+                    IF(IVGTYP(I,J)==LCZ_3) UTYPE_URB=3
+                    IF(IVGTYP(I,J)==LCZ_4) UTYPE_URB=4
+                    IF(IVGTYP(I,J)==LCZ_5) UTYPE_URB=5
+                    IF(IVGTYP(I,J)==LCZ_6) UTYPE_URB=6
+                    IF(IVGTYP(I,J)==LCZ_7) UTYPE_URB=7
+                    IF(IVGTYP(I,J)==LCZ_8) UTYPE_URB=8
+                    IF(IVGTYP(I,J)==LCZ_9) UTYPE_URB=9
+                    IF(IVGTYP(I,J)==LCZ_10) UTYPE_URB=10
+                    IF(IVGTYP(I,J)==LCZ_11) UTYPE_URB=11
+
+
+                    IF(UTYPE_URB==1) FRC_URB2D(I,J)=1.
+                    IF(UTYPE_URB==2) FRC_URB2D(I,J)=0.99
+                    IF(UTYPE_URB==3) FRC_URB2D(I,J)=1.00
+                    IF(UTYPE_URB==4) FRC_URB2D(I,J)=0.65
+                    IF(UTYPE_URB==5) FRC_URB2D(I,J)=0.7
+                    IF(UTYPE_URB==6) FRC_URB2D(I,J)=0.65
+                    IF(UTYPE_URB==7) FRC_URB2D(I,J)=0.3
+                    IF(UTYPE_URB==8) FRC_URB2D(I,J)=0.85
+                    IF(UTYPE_URB==9) FRC_URB2D(I,J)=0.3
+                    IF(UTYPE_URB==10) FRC_URB2D(I,J)=0.55
+                    IF(UTYPE_URB==11) FRC_URB2D(I,J)=1.
+                  ELSE
+                    IF(IVGTYP(I,J)==ISURBAN) UTYPE_URB=2
+                    IF(IVGTYP(I,J)==LCZ_1) UTYPE_URB=1 ! LOW_DENSITY_RESIDENTIAL
+                    IF(IVGTYP(I,J)==LCZ_2) UTYPE_URB=2 ! HIGH_DENSITY_RESIDENTIAL
+                    IF(IVGTYP(I,J)==LCZ_3) UTYPE_URB=3 ! HIGH_INTENSITY_INDUSTRIAL
+
+                    IF(UTYPE_URB==1) FRC_URB2D(I,J)=0.5
+                    IF(UTYPE_URB==2) FRC_URB2D(I,J)=0.9
+                    IF(UTYPE_URB==3) FRC_URB2D(I,J)=0.95
+                  END IF
```

Comment:
I think it is okay to have UTYPE_URB here for those grids that are not dominant urban types (it's better to do urban_var_init). But not for frc_urb2d as the value comes from the geo file, so is different from the input map.

Reply:
@tslin2 Will FRC_URB2D be read or set for grids which aren't dominant urban types if we remove it from here?

Reply:
@epn09 Yes, FRC_URB2D will be read from wrfinput file for all grids including grids that are not dominated by urban types. You could try to remove the FRC_URB2D part here and do a test to see if the model runs fine.

Reply:
@tslin2 I'm not so sure about the default SLUCM, BEP-BEM, and LCZ, but for our new options, I think the `frc_urb2d` statement can be removed if the following
https://github.com/wrf-model/WRF/blob/3cadf04277ac3a050e65461efb6aa939349c60a8/dyn_em/module_initialize_real.F#L3140-L3142
is changed to
```fortran
IF (grid%landusef(i,grid%isurban,j) > 0) THEN 
```
I have tested it but I feel like it should work.

Reply:
> @epn09 Yes, FRC_URB2D will be read from wrfinput file for all grids including grids that are not dominated by urban types. You could try to remove the FRC_URB2D part here and do a test to see if the model runs fine.

yes it work only to set urban type here. In fact, my previous test only removed the FRC_URB2D. I test again today; setting the UTYPE_URB to the exact number is needed, not required to set the FRC_URB2D. This is because a single-layer urban model requires an urban type to read the parameter from the table; an urban type cannot be zero.

Reply:
> @tslin2 I'm not so sure about the default SLUCM, BEP-BEM, and LCZ, but for our new options, I think the `frc_urb2d` statement can be removed if the following
> 
> https://github.com/wrf-model/WRF/blob/3cadf04277ac3a050e65461efb6aa939349c60a8/dyn_em/module_initialize_real.F#L3140-L3142
> 
> 
> is changed to
> ```fortran
> IF (grid%landusef(i,grid%isurban,j) > 0) THEN 
> ```
> 
> I have tested it but I feel like it should work.

the frc_urb2d is not the issue here, but the urban type, UTYPE_URB. If the model has urban type 0, it cannot get the parameter from the urban table, and the model will crash. BEP and BEP-BEM are not yet compatible with mosaic.

Reply:
> @tslin2 Will FRC_URB2D be read or set for grids which aren't dominant urban types if we remove the statements from here?

the FRC_URB2D is first read from the wrfinput file, then set in the urban_var_init.
In urban_var_init, if the dominant land cover type is not urban, the FRC_URB2D is set to zero. 
So for the mosaic case, if we remove the FRC_URB2D here, for the grids that urban is not the dominant land cover type, the FRC_URB2D is zero, which means the impervious fraction is zero.